Creation of a page for versioning (#465)
neo4j/docs-getting-started: modules/ROOT/pages/data-modeling/versioning.adoc

> = Versioning

Reviewer: ```suggestion
= Track change
```

:description: See what options of graph data model versioning are commonly used in combination with Neo4j.

> Every time you xref:data-modeling/graph-model-refactoring.adoc[refactor] your data model, you create new versions of it.

Reviewer: I'm not disputing the meaning of _versioning_, what I mean is what this section contains are various models available for tracking change, not tracking change of the data model.

RETURN p.price
--

.Get the current product catalogue and the prices
[source,cypher]
--
MATCH ()-[r:HAS_PRODUCT]->(p)
WHERE r.validTo = ∞
RETURN p.name, p.price
--

== Linked list

A linked list is also a modeling design that can be particularly useful whenever the sequence of objects matters.
Versioning is a process that uses sequencing to track changes, thus linked lists can be a versioning method too.

Linked lists are particularly useful when you need to know:

* In what order certain changes were made in a data model (e.g. getting the order of transactions executed on a bank account).
* What are the previous and next elements of a list, based on the relationship between them (e.g. what song is the next on a playlist, or undo an action on a text document).

image::linked-list-versioning.svg[Example graph showing a linked list model design being used for versioning,width=400,role=popup]

This example of a linked list versioning shows that:

* The entity `Product` is linked to the first element of the sequence, and can be linked to the last one.
* Similarly to the the xref:#_versioning_of_entities[versioning of entities], the entity `Product` is also immutable here.

> * Each element of the sequence is linked to the next one through a `NEXT` relationship, thus showing a sequence without necessarily numbering versions.

Reviewer: ```suggestion
* Each element of the sequence is linked to the next one through a `NEXT` relationship.
```
This example is using different names/labels instead of numbers, not vastly different.

=== Pros and cons

[cols="<,<",options="header"]
|===
| **Pros**
| **Cons** 

| Efficient by using relationships to get the next/previous element.
| Limited to very specific use cases without using other versioning patterns.

| Simple modeling and maintenance.
| Difficult to find a specific version which is not the first or the last.

| Explicit for end users.
|
|===

=== Query examples

> These are common queries you could make when using the linked list versioning strategy:

Reviewer: ```suggestion
These are examples of common queries that are useful with the linked-list versioning strategy:
```

.Get the next name of the product named “Professional chair”
[source,cypher]
--
MATCH (:State{name: “Professional chair”})-[:NEXT]->(s:State)
RETURN s.name
--

.Get the previous name of the product with the id '1'
[source,cypher]
--
MATCH (:Product {id:1})-[:LAST]->(:State)<-[:NEXT]-(s:State)
RETURN s.name
--

== Timeline tree

> Besides being a xref:data-modeling/modeling-designs.adoc[modeling design] on its own, timeline trees can also be used for versioning,

Reviewer: ```suggestion
As mentioned in xref:data-modeling/modeling-designs.adoc[Modeling designs], the timeline tree is a common modeling design.
It can be a useful strategy when you want to track change.
```
This is also modeling, but specific for tracking change, so it doesn't make sense to me to make this distinction here.

> In this example, the timeline structure spans from years to days, and the rest of the non-time data nodes are the nodes that contain the important pieces data in the graph:

Reviewer: ```suggestion
In this example, the timeline structure spans from years to days, and the rest of the non-time data nodes are the nodes that contain the important pieces of data in the graph:
```

image::timeline-tree.svg[Graph with two different timeslines divided in years, months, and days and what purchases are connected to these dates,width=600,role=popup]

=== Query examples

> If you want to find all purchases that happened in a given time period, such as every purchase in the month of December 2012, the timeline tree can be navigated from 2012, to December, and then fetch everything from the connected leaf nodes (nodes with no descendants) under that branch:

Reviewer: We've mentioned "leaf nodes" before but this is the first time they're defined. Maybe this can be defined earlier on?

[source,cypher]
--
MATCH (root:Timeline)-[:IN_YEAR]->(year:Year {value:2012})-[:IN_MONTH]->(month:Month {value:12})
WITH month
MATCH (month)-[:ON_DAY]->(day)
MATCH (purchase:Purchase)-[:OCCURRED]->(day)
RETURN purchase
--

== Combined approach

Some complex use-cases require the combination of one or more of the previously mentioned modeling techniques since each has advantages and disadvantages.

> The right combination is strongly dependent on the specific use-case. 

Reviewer: ```suggestion
The right combination depends on the specific use-case. 
```

> Factors such as query times and the frequency of transactions should be considered as well.

Reviewer: What is *query times*?
